### PR TITLE
Add Andorra<->Spain exchange

### PR DESCRIPTION
**Pull request by silimotion**
## Issue

Closes partly #2912 

## Description

This PR adds the connection between Spain and Andorra, as discussed in the issue. I've also updated the ESIOS parser to use ZoneKey.

### Preview

Preview of the connection using the mockserver:

Screenshot_20230724_203447

### Double check

- [X] I have tested my parser changes locally with `poetry run test_parser "zone_key"`
- [x] I have run `pnpx prettier --write .` and `poetry run format` to format my changes.

### CODE REVIEW

File: parsers/ESIOS.py [hunk 1]
```diff
@@ -45,7 +45,19 @@ def fetch_exchange(
...
     dates = {"start_date": start_date, "end_date": end_date}
     query = urlencode(dates)
-    url = "https://api.esios.ree.es/indicators/10209?{0}".format(query)
+
+    zone_key = ZoneKey("->".join(sorted([zone_key1, zone_key2])))
+    if zone_key == ZoneKey("AD->ES"):
+        url = "https://api.esios.ree.es/indicators/10278?{0}".format(query)
+    elif zone_key == ZoneKey("ES->MA"):
+        url = "https://api.esios.ree.es/indicators/10209?{0}".format(query)
+    else:
```
Comment (VIKTORVAV99) [inline parsers/ESIOS.py:54]: While this works well for this change I had another approach in mind to make this more extensible and easier to re-use in the future. (for example adding demand parsers for the Spanish islands).

If you create a format_url function (or get_url might be a better name) like this:
```python
def format_url(target_datetime: datetime, ID: str):
    start_date = (target_datetime - timedelta(hours=24)).isoformat()
    end_date = target_datetime.isoformat()
    dates = {"start_date": start_date, "end_date": end_date}
    query = urlencode(dates)
    return f"https://api.esios.ree.es/indicators/{ID}?{query}"
```
And then create a mapping like this (as a constant above the other code):
```python
EXCHANGE_ID_MAP = {
    "AD->ES": "10210",
    "ES->MA": "10209",
}
```
We can call it like this:
```python
    url = format_url(target_datetime, EXCHANGE_ID_MAP[zone_key])
```

This means we can reuse the format_url function and just switch out the mapping if we create new parser functions.

**Comment by silimotion** [comment]
I was wondering, what is gonna happen when Spain imports from Andorra? As there is no parser, will the carbon intensity be unknown, or is there a default somewhere? The zone config doesn't seem to contain any default.

**Comment by VIKTORVAV99** [comment]
> I was wondering, what is gonna happen when Spain imports from Andorra? As there is no parser, will the carbon intensity be unknown, or is there a default somewhere? The zone config doesn't seem to contain any default.

Since there is no fallback for the zone it will use the world average for that year.
I'll try to do a full review of this later today so I can re-fetch all the data as soon as possible.

File: parsers/ESIOS.py [hunk 2]
```diff
@@ -18,10 +18,24 @@
 
 TIMEZONE = pytz.timezone("Europe/Madrid")
 
+# Map each exchange to the ID used in the API. 10278 is used for Andorra->Spain until 10210 returns data correctly
+EXCHANGE_ID_MAP = {
+    "AD->ES": "10278",
```
Comment (VIKTORVAV99) [inline parsers/ESIOS.py:23]:
```suggestion
    "AD->ES": "10278", # Switch to 10210 when it has data.
```

**Review by VIKTORVAV99 — COMMENTED** [review]
Sorry for the delay here but I took some extra time to make sure we can actually use this dataset and as far as I can tell there should be no issues right now.

I have added two small suggestions for you to take a look at but consider this ready to be merged in the current state if you rather not change things.

File: parsers/ESIOS.py [hunk 3]
```diff
@@ -58,12 +73,15 @@ def fetch_exchange(
...
 
     for value in values:
         # Get last value in datasource
         # Datasource negative value is exporting, positive value is importing
-        net_flow = -value["value"]
+        # If Spain is the second country, the value must be opposite
+        if zone_key.startswith("ES"):
+            net_flow = -value["value"]
+        else:
+            net_flow = value["value"]
```
[the remark this answers is not in the document]
Comment (VIKTORVAV99) [reply parsers/ESIOS.py:84]: The parsers should always get called with the keys in alphabetical order so as far as I know this should work.

[the remark this answers is not in the document]
Comment (silimotion) [reply parsers/ESIOS.py:84]: And how about this?:
 `zone_key.startswith("ES->")`

[the remark this answers is not in the document]
Comment (VIKTORVAV99) [reply parsers/ESIOS.py:84]: It works but I'd prefer a strict comparison. But I'll leave that up to you.

**Review by VIKTORVAV99 — APPROVED** [review]

**Comment by VIKTORVAV99** [comment]
Thanks for implementing the changes. I'm merging this now.

**Pull request merged**